### PR TITLE
feat: save assessment questions

### PR DESCRIPTION
**Pull request by joyce-shi**
## Notion ticket link

[Save Assessment Questions section on Create Assessment page](https://www.notion.so/uwblueprintexecs/Save-Assessment-Questions-section-on-Create-Assessment-page-a7843e084ec746ed9cfc2407b05f0e67?pvs=4)

## Implementation description
* Functionality for saving assessment questions

Note: I bumped the Typescript version to resolve an issue I was having defining optional array fields in interfaces, similar to the one described [here](https://github.com/microsoft/TypeScript/issues/41651).

## Steps to test
1. no breaking changes
2. elements of the question editor save properly

### CODE REVIEW

File: frontend/src/APIClients/mutations/TestMutations.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ export const UNARCHIVE_TEST = gql`
   }
 `;
 
-export const SAVE_ASSESSMENT = gql`
+export const CREATE_NEW_ASSESSMENT = gql`
   mutation Create_Test($test: TestRequestDTO!) {
     createTest(test: $test) {
       id
```
Comment (joyce-shi) [inline frontend/src/APIClients/mutations/TestMutations.ts:44]: updated the naming, since this is the same mutation we should use when pressing the **Publish** button on the Create Assessment page

File: frontend/src/components/assessments/assessment-creation/BasicInformation.tsx [hunk 2]
```diff
@@ -17,70 +19,90 @@ import {
...
-                  placeholder=""
+                  placeholder="Select a grade"
+                  selectedOptionStyle="check"
                   useBasicStyles
-                  value={value}
+                  value={
+                    gradeOptions.find(
+                      (option) => option.value === watch("grade"),
+                    ) || undefined
+                  }
```
Comment (joyce-shi) [inline frontend/src/components/assessments/assessment-creation/BasicInformation.tsx:110]: I think we should probably split some of these `Select` components into separate components. I think there's a couple bugs related to dropdowns and also a ticket for refactoring some stuff into common components, so I think I'll work on all of these things together in a subsequent PR! So leaving in this component for now

File: frontend/src/types/QuestionTypes.ts [hunk 3]
```diff
@@ -1,10 +1,41 @@
...
-  MULTIPLE_CHOICE = "Multiple Choice",
-  SHORT_ANSWER = "Short Answer",
-  MULTI_SELECT = "Multi-select",
+  QUESTION_TEXT = "QUESTION_TEXT",
+  TEXT = "TEXT",
+  IMAGE = "IMAGE",
+  MULTIPLE_CHOICE = "MULTIPLE_CHOICE",
+  SHORT_ANSWER = "SHORT_ANSWER",
+  MULTI_SELECT = "MULTI_SELECT",
+}
```
Comment (joyce-shi) [inline frontend/src/types/QuestionTypes.ts:21]: Updated the naming of this values of this enum to match the corresponding backend enum. Then, in all the places where the value was used (i.e as a text label), used our existing formatting utils.

File: frontend/src/types/QuestionTypes.ts [hunk 4]
```diff
@@ -1,10 +1,41 @@
...
+
+export type QuestionElementDataType =
+  // remove string when all the elements have been implemented
```
Comment (joyce-shi) [inline frontend/src/types/QuestionTypes.ts:24]: There's some placeholder values for other elements right now, so leaving this in. Later, when we have all the elements implemented, can remove this line!

File: frontend/src/utils/QuestionUtils.ts [hunk 5]
```diff
@@ -98,7 +106,52 @@ export const getQuestionTexts = (question: QuestionElement[]): string[] => {
...
+};
+
+const formatMultipleChoiceRequest = (
+  data: MultipleChoiceData,
+): MultipleChoiceMetadata => {
+  return {
+    options: data.options.map((option) => option.value),
+    answerIndex: data.options.findIndex((option) => option.isCorrect),
+  };
+};
```
Comment (joyce-shi) [inline frontend/src/utils/QuestionUtils.ts:125]: for other question components, i just stored their `data` in their metadata types. but for multiple choice, since I needed an id field in its implementation, i just defined this extra helper function to convert it to its metadata type!

**Comment by joyce-shi** [comment]
cc @cindy-531

File: package-lock.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (carissa-tang) [inline package-lock.json]: is this change needed ? since we are updating frontend/yarn.lock already

File: frontend/src/components/question-creation/question-elements/modals/multiple-choice/MultipleChoiceModal.tsx [hunk 7]
```diff
@@ -33,7 +33,7 @@ const MultipleChoiceModal = ({
   const [emptyOptionError, setEmptyOptionError] = useState(false);
 
   useEffect(() => {
-    setOptionCount(data ? data.optionCount : 0);
+    setOptionCount(data ? data.options.length : 0);
```
Comment (carissa-tang) [inline frontend/src/components/question-creation/question-elements/modals/multiple-choice/MultipleChoiceModal.tsx]: wait do we still need `optionCount` or can we just use `options.length`

**Review by carissa-tang — COMMENTED** [review]
1. should we add a placeholder here
image
2. can we clear these errors on change
image

File: package-lock.json (hunk 6 above)
Comment (joyce-shi) [reply package-lock.json]: removed!

File: frontend/src/components/question-creation/question-elements/modals/multiple-choice/MultipleChoiceModal.tsx (hunk 7 above)
Comment (joyce-shi) [reply frontend/src/components/question-creation/question-elements/modals/multiple-choice/MultipleChoiceModal.tsx]: i think it still makes sense to keep `optionCount`? bc we use it as a local variable to add elements to `options` (so if we just used `options.length` it would always be 0)

**Comment by joyce-shi** [comment]
addressed pr comments & added handling for multi-select (since jenna was able to get that merged yesterday!!) also removed `package-lock.json` since i think that was accidentally added

File: frontend/src/components/question-creation/question-elements/modals/multiple-choice/MultipleChoiceModal.tsx (hunk 7 above)
Comment (carissa-tang) [reply frontend/src/components/question-creation/question-elements/modals/multiple-choice/MultipleChoiceModal.tsx]: ah kk

**Review by carissa-tang — APPROVED** [review]
LGTM!!
